### DISCUSSION TITLE
implement Zero One Inflated Beta (ZIOB) distribution?

### DISCUSSION THREAD

**Discussion by drbenvincent** [pymc-devs/pymc-extras#19] [Ideas]
PyMC3 has some zero inflated discrete distributions, `ZeroInflatedBinomial`, `ZeroInflatedPoisson`, `ZeroInflatedNegativeBinomial`, but no continuous distributions. One very useful one to have would be the Zero One Inflated Beta (ZIOB) distribution - it would allow better estimation of proportions where there are floor and ceiling effects, similarly it would allow for regression for proportion data (beta regression) where you get these floor and ceiling bound issues.

It looks like it would be a modest job to hand code a model by splitting the data up into floor/ceiling/proportion data and having separate likelihoods, as in this JAGS implementation (see below). Maybe that is the best way to go, but I wanted to stimulate discussion on whether it would be useful for someone to implement a specific ZOIB distribution? Seems like it would be pretty nice from a user/model-builder perspective. It would also be of use to the GLM module presumably

Useful resources for more information
- introductory blog post by Matti Vuorre, [How to analyze visual analog (slider) scale data?](https://vuorre.netlify.app/post/2019/02/18/analyze-analog-scale-ratings-with-zero-one-inflated-beta-models/)
- blog post + JAGS implementation: [Better living through zero-one inflated beta regression](https://mbjoseph.github.io/posts/2018-12-25-better-living-through-zero-one-inflated-beta-regression/)
- [zoib: An R Package for Bayesian Inference for Beta Regression and Zero/One Inflated Beta Regression](https://journal.r-project.org/archive/2015/RJ-2015-019/RJ-2015-019.pdf)
- a discussion on implementation details in the context of GLM's using BRMS in STAN... [Zero-one inflated beta as “censored” beta?](https://discourse.mc-stan.org/t/zero-one-inflated-beta-as-censored-beta/8468/5)

**Comment by ricardoV94** [comment]
I don't know whether a custom distribution makes sense within the primitives of Pymc. This would be a continuous-discrete mixture and usually Pymc distinguishes between the two for sampling and transformation procedures. I am not even sure that pm.Mixture was explicitly designed to allow for mixtures of distributions with different supports. I am just raising questions here, not giving answers by any means.

On the topic itself, I wonder how well/badly a censored normal would deal with this type of data. It avoids both the boundary issues and unlike the ZOIB it would naturally combine both the 0s and 1s into it's mean.

**Comment by drbenvincent** [comment]
I think it would depend on the form of the data and the actual data generating process. For example, you could have data where proportions are tightly clustered around 0.5, but then many more observations at 0 and 1 than you'd expect from the tails of the distribution. 

So a censored normal might do it in some situations, but if the 0's or 1's are actually being generated by a different process from the mean/spread of a normal or beta, then you would need the ZIOB distribution.

**Comment by ricardoV94** [comment]
Yeah, I was just thinking in reply to the blog article. I saw that someone on STAN proposed another model based on two latent ordered cutoffs.

The ZOIB seems to be the most compartmentalized approach, where there are 3 independent processes going on (no dependence between 0s, beta and 1s as far as I can tell). Do you have any intuition on whether this is a standard model by now or if the  "winner" for this type of data is still to be decided?

**Comment by drbenvincent** [comment]
My understanding (which could be wrong) is that there wouldn't be a winner as such, but the censoring and the ZIOB modes would just reflect different ideas about the data generating process.

In my broad area of behavioural science (cognitive science / decision making more specifically) there is a great deal of variation in the level of modelling expertise. On one end of the spectrum, you'll get people running linear regression on censored or truncated data, not even knowing that it's a problem. On the other hand there are more engaged and experienced modellers who do care a lot about getting the most appropriate model, or coming up with custom generative models.

To make things even more interesting, this paper describes use of Tobit (censored) regression, but also what they call a hurdle model which as far as I can tell is basically a ZOIB type model. It's about the dictator game where people choose how much of stuff to share with someone. For example:

_"An equally plausible analytic approach assumes that the decision to [give], and the decision how much to give, conditional on the willingness to give at all, are two separate processes. Then, a hurdle model is appropriate. It first analyses the binary decision to give or not to give with a logit model. For the size of positive contributions, conditional on giving anything, it uses OLS, adjusting the distribution assumption to the fact that observations are taken from a truncated normal distribution (for background see McDowell 2003)."_

> Engel, C. (2011). Dictator games: A meta study. Experimental economics, 14(4), 583-610.

Here's an example distribution of "give amounts" from the paper which might help work out if there is a need for a ZIOB distribution or not:
<img width="699" alt="Screenshot 2021-02-15 at 17 21 34" src="https://user-images.githubusercontent.com/6765047/107976964-46588400-6fb2-11eb-90d1-4b9e5505e611.png">
- The worst approach would basically calculate means on data like this across conditions, or as a function of some variable of interest. This would be pretty hopeless, but people do it.
- Next best would be someone running a censored regression. My guess is most people (in my field) would probably just use some censored data R package.
- Next best would be to at least consider whether this is really censored regression or if it's a ZOIB model. In which case, having a ZIOB distribution could be useful. Making ZOIB more accessible would help nudge people from the more wrong approach to the less wrong approach.
- But then someone going beyond just using packages and writes their own custom models would see the peak at 0.5 and understand that this is actually quite a complex mixture distribution. For such a niche situation, people would really be hand coding their own models.

So on the issue of whether it would be useful to implement a ZIOB likelihood, I think it just comes down to whether it can help nudge people into doing more appropriate analyses. 

On the other side of the argument, you could not bother with a custom likelihood and do a custom ZOIB implementation, as in the R/JAGS blog post link above. That would let people see how to do it, but also make it easier to see how you could extend the approach to account for the peak of people giving exactly 50%, for example.

Sorry about the long post... just thinking through what the best approach is.

**Comment by ricardoV94** [comment]
Thanks for your detailed input. Regardless of what is decided, this would make a great study case for pymc-examples (showing different alternative generative processes for this kind of data)

**Comment by drbenvincent** [comment]
Agreed - when I get a chance I'd quite like to put something together on this
